opengovsg/postmangovsg
# Add package for interpolating the copywriting on frontend

## Problem

Someone who wants to deploy Postman will have to dive into particular parts of the code base to change the copywriting.

Closes #288 

## Solution

Installed and configured [lingui](https://lingui.js.org/tutorials/react.html) to have `en` as the source language and `en-custom` as a "translated" language. User can change the copywriting by providing "translation" for `en-custom` language These copywriting will be displayed when the language is set to `en-custom` (default language is set to `en-custom` in this PR).

**Features**:

- User can change the copywriting offline via translation file without diving into the code.
- Developer can use command line to extract the text to be translated. This is provided by lingui.

## Before & After Screenshots

**BEFORE**:
![image](https://user-images.githubusercontent.com/6182649/86534385-af4f2800-bf0a-11ea-9626-df64dcd0f77a.png)

**AFTER**:
As a proof-of-concept, the translated copywriting is the original copywriting enclosed with square brackets.
![image](https://user-images.githubusercontent.com/6182649/86534373-a0687580-bf0a-11ea-9c0c-cb72a2a4b7ac.png)

## Tests

No automated test was written for this PR.
To confirm functionality, perform manual test on local
1. Install new dependencies and dependencies added in `package.json`
2. Update the copywriting in `frontend/src/locales/en-custom/messages.po`
3. Run `npm run compile`
4. Go to http://localhost:3000/login
5. Verify that the new copywriting is displayed.

## Deploy Notes

**Build**
Need to run `npm run compile` to create runtime catalogs before building the app for production.

Consider to update the build script to include the step.
e.g. in `package.json`, update build script to
```
"scripts": {
  ...
  "build": "npm run compile && INLINE_RUNTIME_CHUNK=false react-scripts build",
  ...
},
```
Need advice on what is the best approach.


**New environment variables**:

Currently the two environment variables `REACT_APP_LOGIN_EMAIL_TEXT` and `REACT_APP_LOGIN_EMAIL_PLACEHOLDER` are not in use (correct me if I am wrong) and are relevant to this issue. We can probably remove them after this PR is merged. 


**New scripts**:

The following commands are added to `scripts` in `package.json`, according to [lingui install instruction](https://lingui.js.org/tutorials/setup-cra.html#install)
- `"add-locale": "lingui add-locale"`: see https://lingui.js.org/ref/cli.html#add-locale
- `"extract": "lingui extract"`: see https://lingui.js.org/ref/cli.html#extract
- `"compile": "lingui compile"`: see https://lingui.js.org/ref/cli.html#compile


**New dependencies**:

New dependencies according to [lingui install instruction](https://lingui.js.org/tutorials/setup-cra.html#install)
- `"@lingui/react": "^2.9.1"` : lingui React components for translations


**New dev dependencies**:

New dev dependencies according to [lingui install instruction](https://lingui.js.org/tutorials/setup-cra.html#install)
- `"@babel/core": "^7.10.4"`: Babel compiler core
- `"@lingui/cli": "^2.9.1"`: lingui CLI for working wit message catalogs
- `"@lingui/macro": "^2.9.1"`: lingui Macro for generating messages in ICU MessageFormat syntax
- `"@types/lingui__core": "^2.7.0"`: TypeScript definitions for @lingui/core
- `"@types/lingui__macro": "^2.7.3"`: TypeScript definitions for @lingui/macro
- `"@types/lingui__react": "^2.8.2"`: TypeScript definitions for @lingui/react
- `"babel-core": "^7.0.0-bridge.0"`: A placeholder package that bridges babel-core to @babel/core
